inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/injection/ClassInjector.java

```diff
@@ -0,0 +1,224 @@
+
+        while (true) {
+            String name = namePrefix + Math.abs(nameGenerationRandom.nextLong());
+            try {
+                Class.forName(name, false, loader);
+            } catch (ClassNotFoundException e) {
+                return name;
+            }
+        }
+    }
+
+    private void markClassForReuse(String classStructureIdentifier, Class<?> clazz) {
+        ClassLoader loader = clazz.getClassLoader();
+        orphanClasses.computeIfAbsent(loader, (cl) -> new HashMap<>())
+                .computeIfAbsent(classStructureIdentifier, (cl) -> new LinkedList<>())
+                .add(new WeakReference<>(clazz));
+    }
+
+    private Optional<Class<?>> tryReusingClassInLoader(String classStructureIdentifier, ClassLoader loader) {
+        if (loader == null) {
+            loader = bootstrapChildLoader;
+        }
+        collectOrphanClasses();
+        LinkedList<WeakReference<Class<?>>> orphansOfGivenStructureInClassloader =
+                orphanClasses.getOrDefault(loader, Collections.emptyMap())
+                        .getOrDefault(classStructureIdentifier, new LinkedList<>());
+
+        WeakReference<Class<?>> classToReuse = orphansOfGivenStructureInClassloader.pollFirst();
+        if (classToReuse == null) {
+            return Optional.empty();
+        } else {
+            return Optional.ofNullable(classToReuse.get());
+        }
+    }
+
+    private void collectOrphanClasses() {
+        Reference<? extends InjectedClass<?>> freedRef = unusedInjectedClassesQueue.poll();
+        while (freedRef != null) {
+            val classRef = (InjectedClassReference) freedRef;
+            activeReferences.remove(classRef);
+            Class<?> injectedClass = classRef.injectedClassObject.get();
+            //the target classloader might have been garbage collected, therefore the null check
+            if (injectedClass != null) {
+                //allow reuse of this class
+                markClassForReuse(classRef.classStructureIdentifier, injectedClass);
+            }
+            freedRef = unusedInjectedClassesQueue.poll();
+        }
+    }
+
+    /**
+     * A specialized {@link WeakReference} to an {@link InjectedClass}, which preserves the actual class object and its structural identifier,
+     * even when the {@link InjectedClass} is garbage collected.
+     */
+    private static class InjectedClassReference extends WeakReference<InjectedClass<?>> {
+
+        private final WeakReference<Class> injectedClassObject;
+        private final String classStructureIdentifier;
+
+        public InjectedClassReference(String classStructureIdentifier, InjectedClass<?> referent, ReferenceQueue<? super InjectedClass<?>> q) {
+            super(referent, q);
+            this.classStructureIdentifier = classStructureIdentifier;
+            Class<?> targetClass = referent.getInjectedClassObject().orElse(null);
+            injectedClassObject = new WeakReference<>(targetClass);
```

Comment:
would this work with `null` as the argument? in the line above there is `orElse(null)`